components/style/stylist.rs

```diff
@@ -1592,37 +1560,130 @@ impl<'a> SelectorVisitor for StylistSelectorVisitor<'a> {
             Origin::Author => &mut self.author,
             Origin::User => &mut self.user,
         }
     }
+
+    fn clear(&mut self) {
+        self.user_agent.clear();
+        self.author.clear();
+        self.user.clear();
+    }
+
+    fn has_rules_for_pseudo(&self, pseudo: &PseudoElement) -> bool {
+        self.iter_origins().any(|d| d.has_rules_for_pseudo(pseudo))
+    }
+
+    fn iter_origins(&self) -> CascadeDataIter {
+        CascadeDataIter {
+            cascade_data: &self,
+            cur: 0,
+        }
+    }
+}
+
+struct CascadeDataIter<'a> {
+    cascade_data: &'a CascadeData,
+    cur: usize,
+}
+
+impl<'a> Iterator for CascadeDataIter<'a> {
+    type Item = &'a PerOriginCascadeData;
+
+    fn next(&mut self) -> Option<&'a PerOriginCascadeData> {
+        let result = match self.cur {
+            0 => &self.cascade_data.user_agent,
+            1 => &self.cascade_data.author,
+            2 => &self.cascade_data.user,
+            _ => return None,
+        };
+        self.cur += 1;
+        Some(result)
+    }
+}
+
+/// Data resulting from performing the CSS cascade that is specific to a given
+/// origin.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Debug)]
+struct PerOriginCascadeData {
+    /// Rules from stylesheets at this `CascadeData`'s origin.
+    element_map: SelectorMap<Rule>,
+
+    /// Rules from stylesheets at this `CascadeData`'s origin that correspond
+    /// to a given pseudo-element.
+    pseudos_map: FnvHashMap<PseudoElement, SelectorMap<Rule>>,
+}
+
+impl PerOriginCascadeData {
+    fn new() -> Self {
+        let mut data = PerOriginCascadeData {
+            element_map: SelectorMap::new(),
+            pseudos_map: Default::default(),
+        };
+        SelectorImpl::each_eagerly_cascaded_pseudo_element(|pseudo| {
+            data.pseudos_map.insert(pseudo, SelectorMap::new());
+        });
+        data
+    }
+
+    #[inline]
+    fn borrow_for_pseudo(&self, pseudo: Option<&PseudoElement>) -> Option<&SelectorMap<Rule>> {
+        match pseudo {
+            Some(pseudo) => self.pseudos_map.get(&pseudo.canonical()),
+            None => Some(&self.element_map),
+        }
+    }
+
+    #[inline]
+    fn borrow_mut_for_pseudo_or_insert(&mut self, pseudo: Option<&PseudoElement>) -> &mut SelectorMap<Rule> {
+        match pseudo {
+            Some(pseudo) => {
+                self.pseudos_map
+                    .entry(pseudo.canonical())
+                    .or_insert_with(SelectorMap::new)
+            }
+            None => &mut self.element_map,
+        }
+    }
+
+    fn clear(&mut self) {
+        self.element_map = SelectorMap::new();
+        self.pseudos_map = Default::default();
+        SelectorImpl::each_eagerly_cascaded_pseudo_element(|pseudo| {
+            self.pseudos_map.insert(pseudo, SelectorMap::new());
+        });
+    }
+
+    fn has_rules_for_pseudo(&self, pseudo: &PseudoElement) -> bool {
+        self.pseudos_map.contains_key(pseudo)
```

Comment:
On another note, we should probably make the pseudos map be an enumerated array, but that's probably ok for now.